metrics-sampler/src/onboarding_metrics_sampler.rs

```diff
@@ -0,0 +1,131 @@
+//! Samples onboarding-related metrics at a fixed interval
+
+use std::time::Duration;
+
+use common::types::tasks::{HistoricalTaskDescription, WalletUpdateType};
+use state::State;
+
+use crate::sampler::AsyncMetricSampler;
+
+/// The name of the sampler for onboarding metrics
+const ONBOARDING_METRICS_SAMPLER_NAME: &str = "onboarding-metrics-sampler";
+/// The interval at which to sample onboarding metrics
+const ONBOARDING_METRICS_SAMPLE_INTERVAL_MS: u64 = 10_000;
+
+/// Metric describing the number of wallets
+const NUM_WALLETS_METRIC: &str = "num_wallets";
+/// Metric describing the number of wallets with at least one deposit
+const NUM_WALLETS_WITH_DEPOSITS_METRIC: &str = "num_wallets_with_deposits";
+/// Metric describing the number of wallets with at least one order placement
+const NUM_WALLETS_WITH_ORDERS_METRIC: &str = "num_wallets_with_orders";
+/// Metric describing the number of wallets with at least one match
+const NUM_WALLETS_WITH_MATCHES_METRIC: &str = "num_wallets_with_matches";
+
+/// Samples onboarding metrics at a fixed interval
+#[derive(Clone)]
+pub struct OnboardingMetricsSampler {
+    /// A handle to the global state
+    state: State,
+}
+
+impl OnboardingMetricsSampler {
+    /// Create a new `OnboardingMetricsSampler`
+    pub fn new(state: State) -> Self {
+        Self { state }
+    }
+}
+
+impl AsyncMetricSampler for OnboardingMetricsSampler {
+    fn name(&self) -> &str {
+        ONBOARDING_METRICS_SAMPLER_NAME
+    }
+
+    fn interval(&self) -> Duration {
+        Duration::from_millis(ONBOARDING_METRICS_SAMPLE_INTERVAL_MS)
+    }
+
+    async fn sample(&self) -> Result<(), String> {
+        // First metric: number of wallets
+        // Second metric: number of wallets w/ at least one deposit
+        // Third metric: number of wallets w/ at least one order placement
+        // Fourth metric: number of wallets w/ at least one match
+        // Latter 3 metrics need task history. So we fetch all wallet IDs, task history
+        // for all wallets, and bucket into sets from there
+
+        let wallets = self.state.get_all_wallets().await?;
+        let num_wallets = wallets.len();
+
+        let mut num_wallets_with_deposits = 0;
+        let mut num_wallets_with_orders = 0;
+        let mut num_wallets_with_matches = 0;
+
+        for wallet in wallets {
```

Comment:
best to break this out into a helper, `process_wallet_onboarding_metrics` or similar